Fixed virtual builds

noseglasses opened the pull request:
This is one of a set of PRs that intent to fix Kaleidoscope virtual builds.

Changes were made to the following repos
* Kaleidoscope
* Kaleidoscope-Bundle-Keyboardio

All PRs must be merged together for everything to work.

See https://github.com/keyboardio/Kaleidoscope-Bundle-Keyboardio/pull/18 for the overall progress.

noseglasses commented on bin/kaleidoscope-builder:
```diff
@@ -58,23 +58,42 @@ build_filenames () {
-    fi
-    export CCACHE_PATH=${COMPILER_PATH}/
-    CCACHE_ENABLE="-prefs compiler.path=${CCACHE_WRAPPER_PATH}/"
+      if ! [ -d "$CCACHE_WRAPPER_PATH" ]; then
```
The following changes make it possible to use ccache for virtual and non-virtual builds in parallel.

noseglasses commented on bin/kaleidoscope-builder:
```diff
@@ -58,23 +58,42 @@ build_filenames () {
 
 firmware_size () {
-    if [ "${BOARD}" = "virtual" ]; then
+    if [ "${ARCH}" = "x86" ]; then
```
Virtual builds are now triggered by defining `ARCH=x86`. The `BOARD` variable is left to specify the actual hardware that is supposed to be simulated.

obra replied:
Can we call it something like x86_simulator or something? I can see some future where we have keyboard firmware running on an embedded x86 :)

noseglasses replied:
I'd even call the architecture `virtual` because who says that we are going to live with `x86` as host platform for ever. You point about `x86` becoming an embedded platform makes a lot of sense.

noseglasses commented on etc/kaleidoscope-builder.conf:
```diff
@@ -12,10 +12,26 @@ LIBRARY="${LIBRARY:-${SKETCH}}"
+    
+    # Set the compiler path for virtual builds (x86)
+    #
+    if [ -z "${COMPILER_PATH}" ]; then
```
For virtual builds we automatically set compiler information, unless predefined.

noseglasses commented on src/kaleidoscope/Hardware.h:
```diff
@@ -29,16 +29,7 @@
- */
-
-/**
- * Forward declaration of the cRGB structure.
```
This forward declaration became obsolete.

noseglasses commented on src/kaleidoscope/bitfields.cpp:
```diff
@@ -44,8 +44,8 @@ bool _BaseBitfield::isBitSetPROGMEM_P(const void *bit_field, uint8_t raw_pos) {
   const uint8_t *bytes = reinterpret_cast<const uint8_t *>(bit_field);
-  uint8_t byte = pgm_read_byte(&(bytes[byte_id]));
-  return byte & (0x1 << bit_pos);
+  uint8_t the_byte = pgm_read_byte(&(bytes[byte_id]));
```
Name collision with type `byte`.

obra replied:
Maybe lift this to its own tiny PR we can merge in now?

noseglasses commented on src/kaleidoscope/bitfields.h:
```diff
@@ -136,7 +136,7 @@ class Bitfield : public internal::_BaseBitfield {
   //
   constexpr bool operator[](uint8_t raw_pos) const {
-    return this->isBitSetPROGMEM_P(raw_pos);
+    return this->isBitSetPROGMEM_P(&bits_, raw_pos);
```
Missing pointer call argument `&bits__`.

obra replied:
Maybe lift this to its own tiny PR we can merge in now?

noseglasses commented on src/kaleidoscope/hardware/keyboardio/Model01.cpp:
```diff
@@ -16,6 +16,7 @@
  */
 
 #ifdef ARDUINO_AVR_MODEL01
+#ifndef ARDUINO_VIRTUAL
```
Make sure that non-virtual keyboard hardwares are not compiled in virtual builds.

noseglasses commented on src/kaleidoscope/hardware/keyboardio/Model01.cpp:
```diff
@@ -31,13 +32,6 @@ bool Model01::isLEDChanged = true;
 keydata_t Model01::leftHandMask;
 keydata_t Model01::rightHandMask;
 
-static constexpr int8_t key_led_map[4][16] PROGMEM = {
```
Moved to the specification header.

noseglasses commented on src/kaleidoscope/hardware/keyboardio/Model01.cpp:
```diff
@@ -122,10 +116,6 @@ void Model01::setCrgbAt(byte row, byte col, cRGB color) {
   setCrgbAt(getLedIndex(row, col), color);
 }
 
-int8_t Model01::getLedIndex(byte row, byte col) {
```
Moved to the specification header.

noseglasses commented on src/kaleidoscope/hardware/keyboardio/Model01.h:
```diff
@@ -20,15 +20,13 @@
-
-#include "kaleidoscope/macro_helpers.h"
-
-#define CRGB(r,g,b) (cRGB){b, g, r}
```
Moved to the specification header.

noseglasses commented on src/kaleidoscope/hardware/keyboardio/Model01.h:
```diff
@@ -100,36 +100,4 @@ class Model01 : public kaleidoscope::Hardware {
 
 #include "kaleidoscope/hardware/key_indexes.h"
 
-#define PER_KEY_DATA_STACKED(dflt,                                    \
```
Moved to the specification header.

noseglasses commented on src/kaleidoscope/key_defs.h:
```diff
@@ -27,7 +27,7 @@
 
 #ifdef ARDUINO_VIRTUAL
-#include "VirtualHID/VirtualHID.h"
+#include "VirtualHID.h"
```
The directory `VirtualHID` was renamed `MultiReport`. `VirtualHID.h` was moved one level up in the directory hierarchy.

noseglasses commented on src/kaleidoscope/plugin/Colormap.cpp:
```diff
@@ -17,9 +17,6 @@
 
 #include <Kaleidoscope-Colormap.h>
-
-#include <EEPROM.h>
```
No more required and causes problems with virtual builds.

obra replied:
Maybe lift this to its own tiny PR we can merge in now?

noseglasses commented on src/kaleidoscope/plugin/Cycle.cpp:
```diff
@@ -48,7 +48,7 @@ void Cycle::replace(uint8_t cycle_size, const Key cycle_steps[]) {
   Key key;
 
-  key.raw = pgm_read_word(cycle_steps + idx);
+  key.raw = pgm_read_word(&cycle_steps[idx].raw);
```
Fixed due to problems with virtual builds.

noseglasses commented on src/kaleidoscope/plugin/FirmwareDump.cpp:
```diff
@@ -15,6 +15,8 @@
  * this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
+#ifndef ARDUINO_VIRTUAL
```
Firmware dump does not make sense for virtual builds and is not possible to be used there anyway.

noseglasses commented on src/kaleidoscope/plugin/LEDEffect-BootAnimation.cpp:
```diff
@@ -60,7 +60,7 @@ EventHandlerResult BootAnimationEffect::afterEachCycle() {
       Key g;
       g.flags = 0;
-      g.keyCode = pgm_read_word(&greeting_[current_index_]);
+      g.keyCode = pgm_read_byte(&greeting_[current_index_]);
```
This was an actual error. Reading a word and assigning it to a byte does not make sense.

obra replied:
Maybe lift this to its own tiny PR we can merge in now?

noseglasses commented on src/kaleidoscope/plugin/Model01-TestMode.cpp:
```diff
@@ -14,6 +14,8 @@
  * this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
+#ifndef ARDUINO_VIRTUAL
```
Is the HardwareTestMode a replacement for Model01-TestMode? It so, it would make sense to replace its use in the stock firmware.

obra replied:
It is. It should.

noseglasses replied:
Are you ok with me preparing a PR that does the replacement in Model01-Firmware?

noseglasses commented on src/kaleidoscope/plugin/OneShot.cpp:
```diff
@@ -124,7 +124,7 @@ EventHandlerResult OneShot::onKeyswitchEvent(Key &mapped_key, byte row, byte col
         if (prev_key_ == mapped_key && isStickable(mapped_key)) {
           if ((Kaleidoscope.millisAtCycleStart() - start_time_) <=
-              ((double_tap_time_out == -1) ? time_out : double_tap_time_out)) {
+              uint16_t((double_tap_time_out == -1) ? time_out : double_tap_time_out)) {
```
Comparing singed and unsigned values can be problematic.

obra replied:
Maybe lift this to its own tiny PR we can merge in now?

noseglasses commented:
Several commits or parts of those have been submitted as individual PRs.

obra commented:
I believe I am. It will want extra testing, though :)

On Mon, Jun 3, 2019 at 2:23 AM noseglasses <notifications@github.com> wrote:

> *@noseglasses* commented on this pull request.
> ------------------------------
>
> In src/kaleidoscope/plugin/Model01-TestMode.cpp
> <https://github.com/keyboardio/Kaleidoscope/pull/650#discussion_r289652737>
> :
>
> > @@ -14,6 +14,8 @@
>   * this program. If not, see <http://www.gnu.org/licenses/>.
>   */
>
> +#ifndef ARDUINO_VIRTUAL
>
> Are you ok with me preparing a PR that does the replacement in
> Model01-Firmware?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/650?email_source=notifications&email_token=AAALC2FVY5Q5L6NAFSZD5YDPYQFYJA5CNFSM4HSCJ3X2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2KN3KI#discussion_r289652737>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2BSB6RQOP777IQAWNLPYQFYJANCNFSM4HSCJ3XQ>
> .
>

noseglasses commented:
Force pushed an update after changes in upstream master.

noseglasses commented:
This PR fails as it requires https://github.com/keyboardio/KeyboardioScanner/pull/18 to be merged first.

noseglasses commented:
Force-pushed an update that reflects the naming change `ARDUINO_VIRTUAL` to `KALEIDOSCOPE_VIRTUAL_BUILD` as discussed [here](https://github.com/keyboardio/KeyboardioHID/pull/55#issuecomment-501775944).

obra commented:
In a few minutes, I'm going to push up to master having cherry-picked -most- of these changes. The bit one I haven't pulled in is b19144f981c07d16b23d76e1b09e7b71057a05cb, the split of the Model 01 definition into logical and physical hardware definitions.  I know we can get to a nice clean way to do this, but I was finding it very hard to review reasonably in the midst of all the other work that @noseglasses has done. (It also needs some updates in light of the merge of LinearAddressing.)

noseglasses commented:
I just rebased the remaining commit.

obra commented:
Cherry-picked these two commits to master:

7d8bfea Let keyIndex-macro use KeyAddr
91a3616 Removed superfluous typedef also provided by KeyAddr.h

obra commented:
Ugh. A reply I wrote to this yesterday and tried to send by email never got through. I've pasted it below. I think a similar reply I wrote a month ago while in China also got swallowed. That one appears to be gone gone gone:

At this point, all that's left in this PR is the split of a keyboard's implementation into things that are:

* Specific to the implementation of the keyboard's electronics 

or

* Specific to the physical and logical layouts of the keyboard's keys and LEDs

The current factoring doesn't feel right to me.

Partially, the additional cognitive load from the added complexity of splitting out the 'spec' feels relatively high. We've doubled the number of device-specific files and added redundancy inside the Arduino boards file. It's now more complicated to figure out where parts of a device's implementation are.

Partially, we've moved "most" of what makes a given keyboard unique off to another file. The Model 01 is definitely one of the more complicated examples we currently support. In the case of a simpler keyboard, we'd end up with a fairly trivial file for the "Hardware" and another file for the layouts.

At the same time, @algernon has been working to migrate a lot of our controller-level  device support into "Driver" files, which is going to further eviscerate the top-level hardware definitions.

I don't really have an answer I know is going to be good.

Maybe we should be flipping the current factoring around, making the data that describes a keyboard the top-level entity and handing it a hardware device implementation + description to use?

Maybe the spec should just be a datastructure defined in the hardware device's header?

Does the problem I'm describing make any sense at all?

noseglasses commented:
TL;DR

I can only see two approaches that enable us to have a generic virtual hardware.

1) Splitting hardware headers in an "electronics" and a "spec" header.
2) Making a single hardware header usable in two different ways by using `#ifdef`s

> and added redundancy inside the Arduino boards file.

I cannot see how to avoid that as `HARDWARE_IMPLEMENTATION` must define the virtual hardware and we can only use `KALEIDOSCOPE_HARDWARE_H` to get that right in order not to change `Kaleidoscope.h`. So the "physical" hardware must be made known to the virtual hardware via an additional command line defined macro.

---

Below you find more thoughts and explanations about the hardware class design.

> Maybe we should be flipping the current factoring around, making the data that describes a keyboard the top-level entity and handing it a hardware device implementation + description to use?

That's not easily possible and the explanation is a bit lengthy.

@algernon's approach uses a combination of inheritance and composition. The composition part makes a lot of sense as there is an obvious A-is-part-of-B relation between a hardware and its "inventory". Plus, different hardwares might share components. The information of what is composed comes from the additional `HardwareInventory` classes that are necessary to make CRTP work.

I really like that approach.

But as this is already quite an advance OOP scenario with different aspects.

We have a hardware A that consists of components B_i. We want to map it to C++ classes with their A-uses-B, A-inherits-from-B or A-contains-B relations. 

To me describing the hardware with its A-consists-of-B_i relations by a composed class where A-contains-B_i makes great sense.

The point where it gets complicated and confusing are the `HardwareInventory` and `Spec` classes.

In contrast to the in the OOP-context conventional hardware class-approach where a class store data and groups methods the `HardwareInventory` and `Spec` serve a mere meta programming purpose of providing information about types and constants. In C++ jargon `HardwareInventory` would be called __traits classes__ that provide information about their template parameter type. The spec classes I introduced to enable virtual hardwares do something similar but without templates.

Traditional classes commonly use traits classes to gain information about types e.g. to know what they are supposed to store or do. Traits classes on the other hand can define traditional classes as typedefs. Thus, there is a non-symmetric two-way relation between those two types of class usage.

Due to this asymmetry of the relation between traits and traditional classes, we cannot simply flip things around as suggested by @obra.

The only way I can see to kind-off flip things around is to have a hardware traits class that typedefs the "electronics" class type. The problem is that this electronics class on the other hand must know the properties defined by the traits class. This would require a solution to avoid redundant definition of properties.

The hardware traits class could typedef the "electronics" class without actually defining it (that this works is demonstrated [here](https://godbolt.org/z/FDt7sl)). Thus it could be used by the virtual hardware. But not defining means that we must prevent the "electronics" specific dependency headers from being included by the hardware header by enclosing them in `#ifdef...#endif` that would be activated differently by the avr/virtual build systems.

Finally, I'd like to add that the way traits classes are used here together with CRTP to define hardware classes could probably be replaced by a [policy based design](https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design). Both seem equally useful here but the policy based approach could possibly go without the additional traits classes due to the absence of CRTP.

On the other hand, those traits classes are a nice interface to the virtual hardware.

obra commented:
@noseglasses Thanks for the thoughtful explanation.

Let me step back a tiny bit to ask a slightly different question:

What's the spec for the spec class? Exactly what keyboard data or metadata "should" be in the Spec file?

noseglasses commented:
> What's the spec for the spec class? Exactly what keyboard data or metadata "should" be in the Spec file?

The spec headers would need to contain everything that the virtual hardware needs to know about the physical keyboard for being able to simulate it properly. Constants like the keyboard matrix dimensions, number of LEDs of any type and the EEPROM size. Also, any types that are otherwise defined by the hardware classes like cRGB. Although not needed by the virtual hardware class, the `KEY_DATA_*` macros must also be contained.

The problem with the existing hardware headers is that some of them include dependencies that cause problems when build for x86. That's why I came up with the spec headers.

Just when I started writing this, I had another idea about how we could live with a single hardware header.

As our hardware classes are all singletons, there's no reason for them to have any non-static data members. A hardware class could, thus, be a mere facade that operates on a globally defined implementation class that is only defined in its *.cpp file. No implementation details would need to be exposed in the header.

With such an approach we could easily live with a single header for each hardware class and even without any ugly `#ifdef...#endif`s. Most hardware classes, e.g. the Model01 hardware already have a strange mixture of static and non-static data members. We could just make the remaining non-static members static and move everything to the .cpp files. E voila.

This approach would also work with the template classes that @algernon introduced with his recent yet unmerged refactoring. The only change this would require is to divide template classes like the `Hardware` template up into two headers, one for the class definition and one for the implementation. Then we could simply use explicit template instantiation for the actual hardware classes.

[Here](https://godbolt.org/z/A5jkXA)'s a simple example that demonstrates what I mean.

noseglasses commented:
After having played around for a while with a solution that intents to wrap everything behind a hardware class facade, it turns out that such a solution would be possible but much too complicated to solve the simple task of separating specs from implementation details. Also the code of the hardware headers would become even more complex and hard to read than it already is.

Therefore, I suggest to go with a solution that uses good old `#ifdef...#endif` to gain hardware headers that can be used in two different ways. I find, however, the solution with a separate specification-class most readable. Something like this:

```cpp
// Model01.h
#include "kaleidoscope/MatrixAddr.h"

namespace kaleidoscope {
namespace hardware {
namespace keyboardio {

class Model01_Description {

 public:

  static constexpr uint8_t matrix_rows = 4;
  static constexpr uint8_t matrix_columns = 16;
  static constexpr int8_t led_count = 64;

  typedef MatrixAddr<matrix_rows, matrix_columns> KeyAddr;

  struct cRGB {
    uint8_t b;
    uint8_t g;
    uint8_t r;
  };

  static int8_t getLedIndex(KeyAddr key_addr);
};
} // namespace keyboardio

typedef keyboardio::Model01_Description Description;

} // namespace hardware

#ifndef KALEIDOSCOPE_VIRTUAL_BUILD

namespace hardware {
namespace keyboardio {

class Model01 {

  public:

    typedef Model01_Description Description;

    static constexpr uint8_t matrix_rows = Description::matrix_rows;
    ...
};

...

} // namespace keyboardio
} // namespace hardware
#endif // KALEIDOSCOPE_VIRTUAL_BUILD
} // namespace kaleidoscope
```

In this example I used the term __description__ instead of __specification__. Not sure what makes more sense, though.

noseglasses commented:
Just pushed and update. 

* a virtual device was implemented
* the hardware-virtual plugin's content has been moved to Kaleidoscopes virtual device directory which makes the Kaleidoscope-Hardware-Virtual plugin obsolete
* all keyboards have been fixed to support being build as virtual hardware
* all this required some changes in other places, e.g. small modifications of the newly introduced device API

Please keep in mind that virtual builds will only work if changes to the bundle repo are merged as well (https://github.com/keyboardio/Kaleidoscope-Bundle-Keyboardio/pull/18).

noseglasses commented:
Forgot to mention that this PR is based on https://github.com/keyboardio/Kaleidoscope/pull/720

obra commented on src/kaleidoscope/plugin/FocusSerial.h:
```diff
@@ -45,14 +45,14 @@ class FocusSerial : public kaleidoscope::Plugin {
   }
   template <typename Var, typename... Vars>
-  void send(Var v, const Vars&... vars) {
+  void send(Var v, Vars... vars) {
```
This is an example of something small that should probably have just been its own PR.

noseglasses replied:
Right. But it's necessary for the virtual build only (causes problems only on x86 or with recent compilers - not sure).

Do you want me to factor it out?

obra replied:
It's not a big deal and I don't need to throw the extra work on you for this one.

Ideally, PRs should be -small- and easy to review. Big architectural changes that include new features, changed APIs, and updates to every part of the code base to adapt to those new APIs are kind of a worst case for changes. Generally, anything that can be done to factor out changes that support future updates into their own PRs, even if that infrastructure isn't used yet, is a win. This isn't a hard and fast rule, but if functionality can be broken into logical stages, I'd rather review 25 20 line PRs than a 500 line PR. 

Obviously, for example, it doesn't make sense to break apart a PR that updates the codebase to change the calling convention from Key.flags to Key.getFlags(), but if the .raw stuff weren't a breaking change, it wouldn't be obvious to me that we shouldn't have broken apart the change to add the new API into its own PR and then the change to the usages into a second PR.

noseglasses replied:
> Generally, anything that can be done to factor out changes that support future updates into their own PRs, even if that infrastructure isn't used yet, is a win. This isn't a hard and fast rule, but if functionality can be broken into logical stages, I'd rather review 25 20 line PRs than a 500 line PR.

Thanks for clarifying this. I will try to adhere to the proposed workflow with separate PRs wherever possible.

But sometimes things are more complicated when working on larger changes that even require changes to multiple repos.

I understand you point from a reviewer's perspective. But from my personal contributor's perspective I find it more motivating, easier and efficient being allowed to submit larger PRs at least for greater changes. Also, whenever I submit something, I am prepared for it to be refused - which is also part of the game. Once that happens, I do not want to loose too much time that I already have invested on git-fiddling. I am a very fast coder but **very** slow when working with RCS. That's why I often code something, make it work and then submit it to obtain feedback as early as possible, sometimes even before taking too much time for cleaning up the git mess. I found that often to be more efficient because reviewers comments can require also a lot of rebasing work.

My typicall workflow when working on new features and larger changes (new plugin API, transient LED modes, host keymaps, virtual builds, ...) is to first do a lot of coding. I imagine how something works and then I just do it. Often in one go. Mostly with more or less frequent commits that only have backup character rather than being meant for submission. Sometimes ideas and work are evolving on the way. I am not the planing type but rather the lets-take-a-chissel-and-see-what's-inside-the-rock type. With such a proceeding it is extremely tedious and inefficient to plan ahead what to put into which commit. That's why I do not take too much care with commit messages during the creative stage of development. Finally, once work is finished and the implementation is solid, I clean up the mess and try to break down everything into independent commits. That can take many hours of rebasing and cherry picking - which is ok and part of the game. Nothing beats unbranched RCS trees.

This approach leads to PRs like the one I submitted here. It has a well defined aim - enabling virtual builds for all keyboards - and contains a lot of commits, each of which I treat as an independent change and try my very best to describe it as such.

If I would break this PR apart into pieces, it would possibly generate quite some PR interdependencies. Some of these PRs would need to be submitted sequentially. A dependent PR can only be submitted once all of its dependencies have been merged. Otherwise dependent PRs would need to contain commits of other dependee PRs. Submitting PRs sequencially requires extra bookkeeping and constantly keeping track of github merges.

As it is fairly normal for PRs to wait an undefined time in the review pipeline, chances are that I am not available for a while once a dependee PR has been merged. This stalls development until I find time to submit the next dependent PR that is possibly again a dependee of another PR...

Sometimes I am totally busy for a while. In other times I can spend quite a lot of spare time on OSS. In the latter case it feels good to do work, document and submit it as a whole.

@obra, I hope that explains a bit why and how I do what I do and I am open for suggestions about how I could improve my workflow to make both our lives easier?

obra commented:
In general, basing large PRs on other unmerged large PRs is something I try really hard to avoid. It makes it much, much more difficult to review the PR and also means that there's zero chance it will be merged until after the other one gets merged. :/

noseglasses commented:
@obra, in general I agree. 

But the Key redesign was already a required part of the redesign before the device API redesign. That change is necessary as x86 builds are typically done with more recent compiler versions. And those are less tolenant against doing funky things with unions. So the whole virtual build depends upon it.

obra commented:
> But the Key redesign was already a required part of the redesign before the device API redesign. That change is necessary as x86 builds are typically done with more recent compiler versions. And those are less tolenant against doing funky things with unions. So the whole virtual build depends upon it.

If it's actually a precondition, then I withdraw my comment :)

obra commented:
> Thanks for clarifying this. I will try to adhere to the proposed workflow
> with separate PRs wherever possible.
>
> But sometimes things are more complicated when working on larger changes
> that even require changes to multiple repos.
>

Yup. "where possible" is absolutely great.


> I understand you point from a reviewer's perspective. But from my personal
> contributor's perspective I find it more motivating, easier and efficient
> being allowed to submit larger PRs at least for greater changes. Also,
> whenever I submit something, I am prepared for it to be refused - which is
> also part of the game. Once that happens, I do not want to loose too much
> time that I already have invested on git-fiddling. I am a very fast coder
> but *very* slow when working with RCS. That's why I often code something,
> make it work and then submit it to obtain feedback as early as possible,
> sometimes even before taking too much time for cleaning up the git mess. I
> found that often to be more efficient because reviewers comments can
> require also a lot of rebasing work.
>

And I totally understand where you're coming from. Indeed, a lot of my goal
here is to help make sure you don't spend a ton of time on something that's
ultimately going to be a bad fit for the codebase.  We can -always- rework
things later.

For a draft PR of something like the Key rework, what I'd be imagining
might be a flow something like this:

* You do the cool new first-pass implementation of the feature in one commit
* You  convert some sample usages in a critical file or two to show off
what they look like in another commit
* You submit a draft PR for us to work through design issues, without
wasting time on the backcompat work, deprecations, etc.
* We talk about stuff
* Once things look mostly good, then put in the time to do the
deprecations, back-compat, convert all the usages, write the UPGRADING
bits, break things apart as best we can into smaller commits.

I want to be really clear that the goal here is to figure out a good
workflow that helps you write lots of wonderful functionality, spend as
little time as possible messing around with git, and helps get stuff landed
with a minimum of pain and frustration for everybody.

I 100% get that there will be stuff that is only really doable as a giant
rework and needs to be reviewed as such.




> My typicall workflow when working on new features and larger changes (new
> plugin API, transient LED modes, host keymaps, virtual builds, ...) is to
> first do a lot of coding. I imagine how something works and then I just do
> it. Often in one go. Mostly with more or less frequent commits that only
> have backup character rather than being meant for submission. Sometimes
> ideas and work are evolving on the way. I am not the planing type but
> rather the lets-take-a-chissel-and-see-what's-inside-the-rock type. With
> such a proceeding it is extremely tedious and inefficient to plan ahead
> what to put into which commit. That's why I do not take too much care with
> commit messages during the creative stage of development. Finally, once
> work is finished and the implementation is solid, I clean up the mess and
> try to break down everything into independent commits. That can take many
> hours of rebasing and cherry picking - which is ok and part of the game.
> Nothing beats unbranched RCS trees.
>
Yep. I work this way too. I've just spent enough time working with version
control tools and....difficult project leaders that I've gotten pretty
quick at rewriting my history to share.




> This approach leads to PRs like the one I submitted here. It has a well
> defined aim - enabling virtual builds for all keyboards - and contains a
> lot of commits, each of which I treat as an independent change and try my
> very best to describe it as such.
>

Yup. And looking at them, they seem pretty reasonable. It -looks- like the
first couple commits in the PR could each be handed over as a single PR or
cherry-picked without too much pain.


> If I would break this PR apart into pieces, it would possibly generate
> quite some PR interdependencies. Some of these PRs would need to be
> submitted sequentially. A dependent PR can only be submitted once all of
> its dependencies have been merged. Otherwise dependent PRs would need to
> contain commits of other dependee PRs. Submitting PRs sequencially requires
> extra bookkeeping and constantly keeping track of github merges.
>
I don't think it's actually that bad. But I've been wrong before. I'll try
to do a better job of doing commit-by-commit reviews of PRs.


> As it is fairly normal for PRs to wait an undefined time in the review
> pipeline, chances are that I am not available for a while once a dependee
> PR has been merged. This stalls development until I find time to submit the
> next dependent PR that is possibly again a dependee of another PR...
>
> Sometimes I am totally busy for a while. In other times I can spend quite
> a lot of spare time on OSS. In the latter case it feels good to do work,
> document and submit it as a whole.
>
> @obra <https://github.com/obra>, I hope that explains a bit why and how I
> do what I do and I am open for suggestions about how I could improve my
> workflow to make both our lives easier?
>
Much appreciated! I think we do want the same thing. And I do want to do
what I can to make it as easy as I can for you to contribute.
ᐧ

obra commented:
I've cherry-picked the following commits to master:

9540be1
ae43e3b036cc4da79ef033e846ad758ed9d82192
3c7717c90f04109917d3106566e2fd781bdf8df7
fd0e8fba67603800b6f27face2d94fe131439fc1
292a19c4a48674660d5abe2c6c19019ea590de0f

This branch should get rebased on master when you've got a chance. It should end with a dramatically smaller diff :)

noseglasses commented:
> And I do want to do what I can to make it as easy as I can for you to contribute.

Very much a appreciated! And good that we talked about the workflow for larger changes.

obra commented on src/kaleidoscope/device/avr/pins_and_ports.h:
```diff
@@ -123,7 +123,65 @@
 #define PIN_F7 PINDEF(F, 7)
 #endif
-
+#else // ifndef KALEIDOSCOPE_VIRTUAL_BUILD
```
Would it be easy enough to just redefine  PINDEF if KALEIDOSCOPE_VIRTUAL_BUILD is defined?  That'd cut down a lot on the repeated definitions here.

noseglasses replied:
Yep. Just force-pushed an update.

obra commented:
If we're going to have to wrap all of these typedefs in #ifndefs and also will have to explicitly typedef their properties classes, it seems like it might make sense to use macros for that, to hide away the ugliness and to make sure that implementors do the "right" thing consistently.

noseglasses commented:
> it seems like it might make sense to use macros for that, 

Good idea. I revived the `EXPORT_DEVICE` macro that was proposed at an earlier stage of the device API redesign.

noseglasses commented:
Rebased this after #729 got merged. This simplifies things further.

obra commented:
The way I'd like to proceed with this is to continue to chip away at the merge by pulling in things that feel "done" bit by bit. 

I think the next thing on that front is 4873997888a7d0d4e2e7b838cdd39763607b97bd, along with switching to it throughout the codebase.  Right now, the usages are split out across a bunch of commits.

I'm visiting family this week, so may be a bit slower than I'd like, but doing that bit of cherry-picking is certainly within my capabilities unless someone else beats me to it. (@noseglasses I don't want to put you through the hassle of teasing apart the commits, but this is an example of where I might have preferred a different factoring of the commits.)

My eventual goal is to minimize the number of ifdefs we need to guard on behalf of the virtual implementation. The fewer we need, the simpler it will be to make sure that virtualization support is easy to maintain and to add for new hardware.

obra commented:
I have lifted out the EXPORT_DEVICE macro and usages and merged them to master.

obra commented:
Cherry-picked 98825cbbd881bca5dc111b3931ed0d28d631b833 to master.

obra commented on src/kaleidoscope/device/technomancy/Atreus2.h:
```diff
@@ -64,10 +64,12 @@ ATMEGA32U4_KEYBOARD(
 
 }
 
+#ifndef KALEIDOSCOPE_VIRTUAL_BUILD
```
Since this is just a legacy definition that should never be called anymore, does the simulator break without the ifndef? (Can we drop the guard in cases like this?)

noseglasses replied:
By forward declaring the named device types in case of virtual builds, the types can be used and the legacy external references can be defined without `#ifndef` guards. I force-pushed an update.

obra commented:
I'd love to figure out a way to need fewer ifdefs for the simulator going forward. I'd also, I think, like to rename the definition to KALEIDOSCOPE_SIMULATOR. However, this has hung out for long enough, so I'm happy for us to address those in the future. I'm going to merge now.

Thanks for the long and tireless work on this @noseglasses
